subsys/usb/device_next/class/usbd_midi.c

```diff
@@ -0,0 +1,752 @@
+	const struct usbd_midi_config *config = dev->config;
+	struct usbd_midi_data *data = dev->data;
+
+	LOG_DBG("USB-MIDI control to host for %s", dev->name);
+	LOG_DBG("  bmRequestType=%02X bRequest=%02X wValue=%04X wIndex=%04X wLength=%04X",
+		setup->bmRequestType, setup->bRequest, setup->wValue, setup->wIndex,
+		setup->wLength);
+
+	/* midi20 6. Class Specific Command: Group Terminal Blocks Descriptors Request */
+	if (data->altsetting == ALT_USB_MIDI_2 && setup->bRequest == USB_SREQ_GET_DESCRIPTOR) {
+		if (setup->wValue != ((CS_GR_TRM_BLOCK << 8) | ALT_USB_MIDI_2)) {
+			errno = -ENOTSUP;
+			return 0;
+		}
+
+		size_t head_len = config->desc->grptrm_header.bLength;
+		size_t total_len = sys_le16_to_cpu(config->desc->grptrm_header.wTotalLength);
+
+		net_buf_add_mem(buf, (void *) &config->desc->grptrm_header,
+				MIN(head_len, setup->wLength));
+
+		if (setup->wLength > head_len) {
+			net_buf_add_mem(buf, (void *) &config->desc->grptrm_blocks,
+					MIN(total_len, setup->wLength) - head_len);
+		}
+		LOG_HEXDUMP_DBG(buf->data, buf->len, "Control to host");
+	}
+
+	return 0;
+}
+
+static int usbd_midi_class_init(struct usbd_class_data *const class_data)
+{
+	const struct device *dev = usbd_class_get_private(class_data);
+
+	LOG_DBG("Init USB-MIDI device class for %s", dev->name);
+
+	return 0;
+}
+
+static void *usbd_midi_class_get_desc(struct usbd_class_data *const class_data,
+				      const enum usbd_speed speed)
+{
+	const struct device *dev = usbd_class_get_private(class_data);
+	const struct usbd_midi_config *config = dev->config;
+
+	LOG_DBG("Get descriptors for %s", dev->name);
+
+	return (speed == USBD_SPEED_HS) ? config->hs_descs : config->fs_descs;
+}
+
+
+static struct usbd_class_api usbd_midi_class_api = {
+	.request = usbd_midi_class_request,
+	.update = usbd_midi_class_update,
+	.enable = usbd_midi_class_enable,
+	.disable = usbd_midi_class_disable,
+	.suspended = usbd_midi_class_suspended,
+	.resumed = usbd_midi_class_resumed,
+	.control_to_host = usbd_midi_class_cth,
+	.init = usbd_midi_class_init,
+	.get_desc = usbd_midi_class_get_desc,
+};
+
+static struct net_buf *usbd_midi_buf_alloc(uint8_t ep)
+{
+	struct udc_buf_info *info;
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&usbd_midi_buf_pool, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	info = udc_get_buf_info(buf);
+	info->ep = ep;
+
+	return buf;
+}
+
+static inline uint8_t usbd_midi_get_ep(struct usbd_class_data *const class_data,
+				       uint8_t direction)
+{
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(class_data);
+	const struct device *dev = usbd_class_get_private(class_data);
+	const struct usbd_midi_config *cfg = dev->config;
+
+	return (usbd_bus_speed(uds_ctx) == USBD_SPEED_HS)
+		? (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_hs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_hs.bEndpointAddress
+		: (direction == USB_EP_DIR_IN)
+			? cfg->desc->if1_1_in_ep_fs.bEndpointAddress
+			: cfg->desc->if1_1_out_ep_fs.bEndpointAddress;
+}
```

Comment:
Must be like https://github.com/zephyrproject-rtos/zephyr/blob/2a04abafa0241e77956d11c5df0407e3b3fe7aec/subsys/usb/device_next/class/usbd_cdc_ncm.c#L245-L271

This construction offers literally no advantages other than unreadability. Please let the compiler to be smart ass. In general, the code should follow the style of existing class/function implementations.